Step3

hyeongyeong opened the pull request.

hyeongyeong committed:
docs : 요구사항3 기본 요건 정리

hyeongyeong committed:
feat : columnInfo 일급 컬렉선 생성, 쿼리 생성 위임

hyeongyeong committed:
feat : enum 이름 변경

hyeongyeong committed:
refactor : 패키지 이동

hyeongyeong committed:
feat : StringBuilder -> string format으로 변경

hyeongyeong committed:
feat : insertQuery 생성 기능 추가

hyeongyeong committed:
feat : test Dummy class 생성

hyeongyeong committed:
feat : ColumnFields 함수 이름 통일

hyeongyeong committed:
feat : findByIds 구현

hyeongyeong committed:
feat : deleteById 구현

jongmin4943 commented on README.md:
```diff
@@ -26,3 +26,12 @@
   - [X] @Transient 이 포함된 컬럼은 컬럼 생성하지 않음
 - [X] Person Class의 Drop Query 생성
+
+## 3단계 QueryBuilderDML
+
+### 요구사항
+
+- [] Person class 의 insert query 생성
+- [] Person class 의 findAll query 생성
+- [] Person class 의 findById query 생성
+- [] Person class 의 delete query 생성
```
요구사항 정리쪽을 깜빡하신 것 같아요 😅

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnFields.java:
```diff
@@ -0,0 +1,54 @@
+        this.columnFields = extract(clazz);
+    }
+
+    private List<ColumnField> extract(Class<?> clazz) {
+        return Arrays.stream(clazz.getDeclaredFields()).map(ColumnField::new)
+                .filter(ColumnField::isNotTransient).collect(Collectors.toList());
+    }
```
```suggestion
    private List<ColumnField> extract(Class<?> clazz) {
        return Arrays.stream(clazz.getDeclaredFields())
                .filter(field -> !field.isAnnotationPresent(Transient.class))
                .map(ColumnField::new)
                .collect(Collectors.toList());
    }
```

- 개인적으로 Stream 을 쓸땐 체이닝 별로 내려써주시면 읽기 좀 더 편하다고 느껴요 :)
- 만들때 처음부터 Transient 를 제거한다면, 필요없는 객체를 생성하는 비용을 아낄 수 있지 않을까요?

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnFields.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ColumnFields {
+
+    private List<ColumnField> columnFields;
```
일급컬렉션 💯 필드를 final 로 만들면 불변을 더 보장할 수 있을 것 같아요 👍

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnFields.java:
```diff
@@ -0,0 +1,54 @@
+                .filter(ColumnField::isNotTransient).collect(Collectors.toList());
+    }
+
+    public String generatePrimaryKeyQuery() {
+        StringBuilder sb = new StringBuilder();
+        List<ColumnField> primaryKey = columnFields.stream().filter(ColumnField::isPrimaryKey).collect(Collectors.toList());
+        if(primaryKey.isEmpty()) {
+            throw new IllegalArgumentException("Entity에 Id로 정의된 column이 존재하지 않습니다.");
+        }
+        sb.append("primary key (");
+        sb.append(primaryKey.stream().map(ColumnField::getName).collect(Collectors.joining(", ")));
+        sb.append(")");
+        return sb.toString();
+    }
```
기존에 `QueryBuilder` 가 가진 책임이 그대로 넘어왔네요.
제가 의도 했던 방향은 `ColumnField` 와 `ColumnFields` 는 쿼리를 만드는걸 모르고, 단순히 컬럼필드에 대한 정보를 가지고 제공하는 책임을 가지고, `QueryBuilder` 가 쿼리를 만드는 책임을 가지는 방향이었어요. 😢

만약 쿼리를 만드는 로직이 바뀐다면 `QueryBuilder` 와 `ColumnField` 도 바뀌게 되는데 그건 단일책임원칙(SRP) 를 위배한다고 느끼는데 어떻게 생각하시나요? :)

jongmin4943 commented on src/main/java/persistence/sql/meta/TableInfo.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.Objects;
+
+public class TableInfo {
+    private String tableName;
```
```suggestion
public class TableInfo {
    private final String tableName;
```
불변을 지킨다면 어떤 장점을 가질 수 있을까요?

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnField.java:
```diff
@@ -0,0 +1,65 @@
+        return Objects.isNull(columnAnnotation) || columnAnnotation.name().isBlank() ? field.getName() : columnAnnotation.name();
+    }
+
+    public H2ColumnType getColumnType() {
+        return H2ColumnType.of(field.getType());
+    }
```
`H2ColumnType` 를 static 하게 접근하는 모양이네요. 만약 런타임에 DB 를 특정짓게 되면 문제가 될 것 같은데 어떻게 해결할 수 있을까요?

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnField.java:
```diff
@@ -0,0 +1,65 @@
+        return !field.isAnnotationPresent(Transient.class);
+    }
+
+    public String generateWhereClause(Object object) throws Exception {
+        field.setAccessible(true);
+        return String.format("%s=%s", getName(), field.get(object));
+    }
```
`WhereClause` 를 만드는건 누가 가져야할 책임일지 고민해보시면 좋을 것 같아요 👍

jongmin4943 replied:
아직 `ColumnField` 에  쿼리를 만드는 책임이 남아 있어요 😭

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnField.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class ColumnField {
+    private Field field;
```
`Field` 는 런타임 환경에서 바뀌지 않는 값이라 매 실행마다 같은 `Field` 에 대한 연산은 같은 값을 반환해요.
그렇다면, 특정 클래스가 `Field` 를 가지고 필요 정보들을 미리 추출해서 가지고 있다가, 누군가 특정 값을 원하면 건내주는 역할을 해보면 어떨까요?

그럼 매번 `Field` 에 대해 계산하지 않아도 되고, 객체에게 책임을 부여 해볼 수 있을 것 같아요 :)

jongmin4943 commented on src/main/java/persistence/sql/dml/QueryBuilderDML.java:
```diff
@@ -0,0 +1,78 @@
+        return whereClause(String.format(findAllFormat, tableInfo.getTableName()), object);
+    }
+
+    private String columnsClause(Class<?> clazz) {
+        ColumnFields columnInfos = new ColumnFields(clazz);
+        return columnInfos.getColumnFields()
+                .stream().map(ColumnField::getName)
+                .collect(Collectors.joining(", "));
+    }
```
이왕 일급컬렉션으로 만들어주셨으니, 객체에게 시켜보시는건 어떨까요?
[묻지 말고 시켜라(TDA)](https://incheol-jung.gitbook.io/docs/study/object/2020-03-10-object-chap6) 를 참고해보셔도 좋을 것 같아요 👍

jongmin4943 commented on src/main/java/persistence/sql/dml/QueryBuilderDML.java:
```diff
@@ -0,0 +1,78 @@
+
+    private String whereClause(String selectQuery, Object object) throws Exception {
+        String whereQuery = "%s where %s";
+        List<ColumnField> primaryFields = new ColumnFields(object.getClass())
+                .getColumnFields().stream().filter(ColumnField::isPrimaryKey).collect(Collectors.toList());
```
`columnFields.getPrimaryKey` 와 같은 메서드가 있으면 좋을 것 같아요 :)

jongmin4943 commented on src/main/java/persistence/sql/dml/QueryBuilderDML.java:
```diff
@@ -0,0 +1,78 @@
+        return String.format(whereQuery, selectQuery, whereClauses.stream().collect(Collectors.joining(" and ")));
+    }
+
+    private String valueClause(Object object) throws Exception {
+        ColumnFields columnInfos = new ColumnFields(object.getClass());
+        List<Field> fields = columnInfos.getColumnFields().stream()
+                .filter(ColumnField::isNotTransient)
+                .map(ColumnField::getField)
+                .collect(Collectors.toList());
+
+        List<String> values = new ArrayList<>();
+        for (Field field : fields) {
+            field.setAccessible(true);
+            values.add(String.valueOf(field.get(object)));
+        }
```
`ColumnFields` 를 만들때 Transient 를 이미 필터링 하는데 여기서 또 해야하는 이유가 있을까요?

Transient 를 필터링하고, field 의 속성을 만지고 하는등의 작업이 `QueryBuilder` 에서 가져도 되는 책임인지 고민해보시면 좋을 것 같아요 :)

jongmin4943 commented on src/test/java/persistence/sql/dml/QueryBuilderDMLTest.java:
```diff
@@ -0,0 +1,162 @@
+        assertThat(queryBuilderDML.deleteById(person)).isEqualTo(expected);
+    }
+
+    @Test
+    @DisplayName("입력받은 object의 deleteById query 생성 - multiple id 일 때")
+    void generateDelteByIdQuery_multipleIds() throws Exception {
+        DummyPerson2 person = getCustomPerson2();
+
+        QueryBuilderDML queryBuilderDML = QueryBuilderDML.getInstance();
+        String expected = "delete from users where id=1 and nick_name=hyeongyeong";
+        assertThat(queryBuilderDML.deleteById(person)).isEqualTo(expected);
+    }
```
요구사항별로 테스트코드를 잘 짜주셨네요 💯 

다만, 만들어주신 `ColumnField`, `ColumnFields`, `TableInfo` 에 관한 테스트가 없는게 아쉽네요 😭
각각 만들면서 배우는 JPA 과정에서 끝까지 가져갈 클래스들이니 테스트가 필수일 것 같아요 :)

jongmin4943 commented on src/main/java/persistence/Application.java:
```diff
@@ -5,7 +5,7 @@
+import persistence.sql.sample.Person;
 import persistence.sql.ddl.QueryBuilderDDL;
 
 public class Application {
```
실제 DB 환경에서 만드신 쿼리가 잘 작동하는지 테스트가 없네요. 추가해주시면 감사하겠습니다 🙇

jongmin4943 reviewed (changes requested):
안녕하세요 현경님!

3단계 빠르게 진행해주셨네요 👍 코드가 조금씩 개선되어 나가는게 느껴지네요!
좀 더 개선 해볼만한 코멘트들 남겼는데 확인 후 재 요청부탁드려요 🙏

궁금하시거나 고민 되시는 부분이 있으시다면 언제든 코멘트나 DM 주세요! 🔥

hyeongyeong committed:
feat : 불변 객체 -> 처음 할당된 변수의 무결성 보장

hyeongyeong committed:
feat : 책임 분리

- Column : 임의의 Class가 가진 각각의 field 속성을 관리

- ColumnFields : Column의 일급 객체 어떤 class의 Field들의 연산을 담당

- QueryBuidlerDDL : Class를 입력받아 쿼리를 생성

hyeongyeong committed:
feat : Test Code 추가 - TableInfo

hyeongyeong committed:
feat : Test Code 추가 - ColumnField

hyeongyeong committed:
feat : Test Code 추가 - ColumnFields

hyeongyeong committed:
feat : 요구사항 표시

hyeongyeong commented:
아직 구현하지 못한 사항중에 인터페이스화가 있는데
고민하면서 적용중입니다.
일단 현재까지 반영한 내용 PR 드립니다!

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnField.java:
```diff
@@ -0,0 +1,65 @@
+        return options;
+    }
+
+    public String getName(){
+        final var columnAnnotation = field.getAnnotation(Column.class);
+        return Objects.isNull(columnAnnotation) || columnAnnotation.name().isBlank() ? field.getName() : columnAnnotation.name();
+    }
```
개인적으로 이 코드의 조건식을 한번에 이해하기는 어렵다고 느껴져요 😅

저는 간단한 조건문 같은 경우에는 괜찮지만, 코드를 읽을 때 한번에 읽어야 할 양이 늘어나기 때문에 삼항연산자를 지양하고 있어요.

저는 코드를 글쓰기라 생각해요. 한 문장의 길이가 너무 길면 전체 컨텍스트를 이해하는데 더 노력해야하고, 한 호흡에 읽기 힘들면 가독성이 떨어지기 시작하죠.

그런 의미에서 저와 함께 미션 진행할때는 삼항연산자를 쓰지 않고 진행해보시는걸 권장 드릴게요 👍

jongmin4943 commented on src/main/java/persistence/sql/dml/QueryBuilderDML.java:
```diff
@@ -0,0 +1,66 @@
+        return new ColumnFields(object.getClass()).valueClause(object);
+    }
+
+    private String whereClause(String selectQuery, Object object) {
+        String whereQuery = "%s where %s";
+        List<ColumnField> primaryFields = new ColumnFields(object.getClass()).getPrimary();
+
+        String primaryWhereClause = primaryFields.stream().map(columnField -> {
+            try {
+                return columnField.generateWhereClause(object);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }).collect(Collectors.joining(" and "));
+
+        return String.format(whereQuery, selectQuery, primaryWhereClause);
+    }
```
whereClause 를 만드는 책임을 여기로 바꿔주셨군요.

만드실때 try catch 를 해주신 이유가 궁금해요 👀

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnFields.java:
```diff
@@ -0,0 +1,58 @@
+                .collect(Collectors.toList());
+    }
+
+    public List<ColumnField> getColumnFields() {
+        return columnFields;
+    }
```
이왕 일급컬렉션을 만들어주셨는데 불변을 지킬 수 있으면 좋을 것 같아요.
개인적으로 일급컬렉션에 책임을 더 부여하는걸 선호하지만, getter 를 통해 외부로 노출해야할땐 불변 리스트로 반환 해볼 수 있어요 😃 
```suggestion
    public List<ColumnField> getColumnFields() {
        return Collections.unmodifiableList(columnFields);
    }
```
불변을 지키면 어떤 장점이 있을까요?

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnFields.java:
```diff
@@ -0,0 +1,58 @@
+                .stream().map(ColumnField::getName).collect(Collectors.toList());
+    }
+
+    public List<ColumnField> getPrimary() {
+        if(columnFields.stream().noneMatch(ColumnField::isPrimaryKey)) {
+            throw new IllegalArgumentException("Entity에 Id로 정의된 column이 존재하지 않습니다.");
+        }
+
+        return columnFields.stream().filter(ColumnField::isPrimaryKey).collect(Collectors.toList());
+    }
```
실제 [JPA spec](https://download.oracle.com/otn-pub/jcp/persistence-2_2-mrel-spec/JavaPersistence.pdf) 에서는 Entity 에 pk 필드가 존재하지 않으면 에러가 나도록 되어있어요. `ColumnFields` 자체를 생성할때 유효성 검사를 해보면 어떨까요? :)
![image](https://github.com/user-attachments/assets/1acd4b62-f031-4c6d-b2de-22abd5640154)

jongmin4943 commented on src/main/java/persistence/sql/meta/ColumnFields.java:
```diff
@@ -0,0 +1,58 @@
+        return columnFields.stream().filter(ColumnField::isPrimaryKey).collect(Collectors.toList());
+    }
+
+    public String getColumnClause() {
+        return getDeclaredFieldNames().stream()
+                .collect(Collectors.joining(", "));
+    }
+
+    public String valueClause(Object object) {
+        List<Field> fields = columnFields.stream().map(ColumnField::getField).collect(Collectors.toList());
+        fields.forEach(field -> field.setAccessible(true));
+        return fields.stream().map(field-> {
+            try {
+                return String.valueOf(field.get(object));
+            } catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }).collect(Collectors.joining(", "));
+    }
```
`ColumnClause` 와 `valueClause` 를 만드는건 쿼리를 생성하는 책임에 가깝지 않을까요? 🤔

jongmin4943 commented on src/test/java/persistence/sql/meta/ColumnFieldTest.java:
```diff
@@ -0,0 +1,84 @@
+                .isEqualTo(expected);
+    }
+
+    @Test
+    @DisplayName("임의의 Class의 Primary Column 가져오기")
+    void getPrimary() {
+        List<String> expected = Arrays.asList("id");
+        assertThat(new ColumnFields(DummyPerson.class).getPrimary().stream().map(ColumnField::getName)).isEqualTo(expected);
+    }
```
`ColumnFieldTest` 에서 `ColumnFields` 를 테스트하고 있는데 의도하신걸까요?

jongmin4943 commented on src/test/java/persistence/sql/meta/ColumnFieldTest.java:
```diff
@@ -0,0 +1,84 @@
+        assertThat(new ColumnFields(DummyPerson.class).getPrimary().stream().map(ColumnField::getName)).isEqualTo(expected);
+    }
+
+    @Test
+    @DisplayName("임의의 Class의 Primary Column 가져오기 - id가 여러개 정의됨")
+    void getPrimary_multiple() {
+        List<String> expected = Arrays.asList("id", "nick_name");
+        assertThat(new ColumnFields(DummyPerson2.class).getPrimary().stream().map(ColumnField::getName)).isEqualTo(expected);
+    }
```
테스트도 가독성을 위해 bdd(given-when-then)을 이용해보시는건 어떨까요?
```suggestion
    @Test
    @DisplayName("임의의 Class의 Primary Column 가져오기 - id가 여러개 정의됨")
    void getPrimary_multiple() {
        // given
        final ColumnFields columnFields = new ColumnFields(DummyPerson2.class);
        // when
        final List<String> actual = columnFields.getPrimary().stream().map(ColumnField::getName).toList();
        // then
        assertThat(actual).containsExactly("id", "nick_name");
    }
```
개인적으로 여기서 `columnFields.getPrimaryNames` 와 같은 메서드를 추가해 객체에게 메시지를 전달하는것을 지향합니다 :)

jongmin4943 commented on src/test/java/persistence/sql/meta/TableInfoTest.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+
+    @Test
+    @DisplayName("임의의 class에서 테이블 Entity 이름 가져오기")
+    void getTableEntityName() {
+        assertThat(new TableInfo(Dummy.class).getTableName()).isEqualTo("dummy");
+    }
```
테스트를 추가해주셨네요 💯 
다만, 예외에 대한 분기가 있었는데 해당 테스트가 없네요 😢

개인적으로 저는 성공 시나리오 테스트도 중요하지만, 실패 시나리오를 테스트하는것도 매우 중요하다 생각해요 👍

jongmin4943 reviewed (approved):
안녕하세요 현경님!

3단계 피드백 잘 반영 해주셨네요 👍
몇 가지 좀 더 개선 해볼만한 부분에 코멘트 달아 보았습니다! 
이번 단계는 여기서 마무리하겠습니다. 다음단계 진행하며 피드백 확인 부탁드립니다. 🙇

궁금하시거나 고민 되시는 부분이 있으시다면 언제든 코멘트나 DM 주세요! 🔥
